cpp/src/arrow/util/spaced.h

```diff
@@ -0,0 +1,199 @@
+
+#include "arrow/util/align_util.h"
+#include "arrow/util/bit_block_counter.h"
+#include "arrow/util/bitmap_reader.h"
+
+namespace arrow {
+namespace util {
+namespace internal {
+
+/// \brief Compress the buffer to spaced, excluding the null entries.
+///
+/// \param[in] src the source buffer
+/// \param[in] num_values the size of source buffer
+/// \param[in] valid_bits bitmap data indicating position of valid slots
+/// \param[in] valid_bits_offset offset into valid_bits
+/// \param[out] output the output buffer spaced
+/// \return The size of spaced buffer.
+template <typename T>
+inline int SpacedCompress(const T* src, int num_values, const uint8_t* valid_bits,
+                          int64_t valid_bits_offset, T* output) {
+  int num_valid_values = 0;
+  int idx_src = 0;
+
+  const auto p =
+      arrow::internal::BitmapWordAlign<1>(valid_bits, valid_bits_offset, num_values);
+  // First handle the leading bits
+  const int leading_bits = static_cast<int>(p.leading_bits);
+  while (idx_src < leading_bits) {
+    if (BitUtil::GetBit(valid_bits, valid_bits_offset)) {
+      output[num_valid_values] = src[idx_src];
+      num_valid_values++;
+    }
+    idx_src++;
+    valid_bits_offset++;
+  }
+
+  // The aligned parts scaned with BitBlockCounter
+  arrow::internal::BitBlockCounter data_counter(valid_bits, valid_bits_offset,
+                                                num_values - leading_bits);
+  auto current_block = data_counter.NextWord();
+  while (idx_src < num_values) {
+    if (current_block.AllSet()) {  // All true values
+      int run_length = 0;
+      // Scan forward until a block that has some false values (or the end)
+      while (current_block.length > 0 && current_block.AllSet()) {
+        run_length += current_block.length;
+        current_block = data_counter.NextWord();
+      }
+      // Fill all valid values of this scan
+      std::memcpy(&output[num_valid_values], &src[idx_src], run_length * sizeof(T));
+      num_valid_values += run_length;
+      idx_src += run_length;
+      valid_bits_offset += run_length;
+      // The current_block already computed, advance to next loop
+      continue;
+    } else if (!current_block.NoneSet()) {  // Some values are null
+      arrow::internal::BitmapReader valid_bits_reader(valid_bits, valid_bits_offset,
+                                                      current_block.length);
+      for (int64_t i = 0; i < current_block.length; i++) {
+        if (valid_bits_reader.IsSet()) {
+          output[num_valid_values] = src[idx_src];
+          num_valid_values++;
+        }
+        idx_src++;
+        valid_bits_reader.Next();
+      }
+      valid_bits_offset += current_block.length;
+    } else {  // All null values
+      idx_src += current_block.length;
+      valid_bits_offset += current_block.length;
+    }
+    current_block = data_counter.NextWord();
+  }
+
+  return num_valid_values;
+}
+
+/// \brief Expand the buffer but leave spaces for null entries.
```

Comment:
Maybe "Relocate values in buffer into positions of non-null values as indicated by a validity bitmap"?

Reply:
done, 3x